tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1031 @@
+   * @return a List of Operations to update the metric state
+   */
+  @Override
+  @SuppressWarnings("unchecked")
+  public List<Op> updateStateList(
+      Operand<? extends TNumber> labels,
+      Operand<? extends TNumber> predictions,
+      Operand<? extends TNumber> sampleWeights) {
+    Ops tf = getTF();
+    Operand<T> tLabels = cast(tf, labels, type);
+    Operand<T> tPredictions = cast(tf, predictions, type);
+    Operand<T> tSampleWeights = sampleWeights != null ? cast(tf, sampleWeights, type) : null;
+    List<Op> updateOperations = new ArrayList<>();
+    Map<ConfusionMatrixEnum, Assign<T>> varInitializers = Collections.EMPTY_MAP;
+    if (!initialized) {
+      varInitializers = build(tPredictions.shape());
+    }
+    if (isMultiLabel() || getLabelWeights() != null) {
+      // labels should have shape (number of examples, number of labels).
+      List<SymbolicShape<? extends TNumber>> symbols = new ArrayList<>();
+      symbols.add(new SymbolicShape<>(tLabels, "N", "L"));
+      if (isMultiLabel()) {
+        // TP, TN, FP, and FN should all have shape
+        // (number of thresholds, number of labels).
+        symbols.add(new SymbolicShape<>(truePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falsePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(trueNegatives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falseNegatives, "T", "L"));
+      }
+      if (getLabelWeights() != null) {
+        symbols.add(new SymbolicShape<>(getLabelWeights(), "L", ""));
+      }
+      updateOperations.addAll(
+          MetricsHelper.assertShapes(tf, symbols, "Number of labels is not consistent."));
+    }
+
+    Map<ConfusionMatrixEnum, Variable<T>> confusionMatrix = new HashMap<>();
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_POSITIVES, truePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_POSITIVES, falsePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_NEGATIVES, trueNegatives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_NEGATIVES, falseNegatives);
+
+    // Only forward labelWeights to update_confusion_matrix_variables when
+    // multiLabel is false. Otherwise the averaging of individual label AUCs is
+    // handled in AUC.result
+    updateOperations.addAll(
+        MetricsHelper.updateConfusionMatrixVariables(
+            tf,
+            confusionMatrix,
+            varInitializers,
+            tLabels,
+            tPredictions,
+            tf.constant(thresholds),
+            null,
+            null,
+            tSampleWeights,
+            isMultiLabel(),
+            isMultiLabel() ? null : getLabelWeights()));
+    return updateOperations;
+  }
+
+  /**
+   * Interpolation formula inspired by section 4 of Davis & Goadrich 2006.
+   *
+   * @return an approximation of the area under the P-R curve.
+   */
+  private Operand<T> interpolatePRAuc() {
+    // truePositives[:self.numThresholds - 1]
+    Ops tf = getTF();
+    Operand<T> tp0 =
```

Comment:
What would you think about building up some notation ahead of time to make the heart of this code more readable? For example:
~~~Java
private Operand<T> positive(Operand<T> x) {
  return tf.math.maximum(x, tf.dtypes.cast(tf.constant(0), x.type()));
}

private Operand<T> isPositive(Operand<T> x) {
  return tf.math.greater(x, tf.dtypes.cast(tf.constant(0), x.type()));
}

private Operand<T> slice(Operand<T> x, int begin, int size) {
  return tf.slice(x, tf.constant(new int[] {begin}), tf.constant(new int[] {size})));
}

private Operand<T> interpolatePRAuc() {
  Ops tf = getTF();
  Operand<T> tp0 = slice(truePositives, 0, getNumThresholds() - 1);
  Operand<T> tp1 = ...
  ...
  Operand<T> p0 = tf.math.add(tp0, fp0);
  Operand<T> p1 = ...

Reply:
Sounds good, but should this be in a util class somewhere  rather than private to AUC?

Reply:
I have added them as AUC private methods for now.

Reply:
👍 done (although I'll raise specific issues in the new code)
I raised #246 for broadly addressing operation API usability.